db: initial work on tenant restraints

dcramer opened the pull request:
putting this up for initial feedback on the approach

still a lot to be done even to constrain project (which would be the first PR)

getsentry-danger commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

dcramer commented:
found more endpoints which need hooked -- still no obvious universal way to do it that doesnt require override a few various implementations

dcramer commented:
The major deciding thing yet is how these should interact with ``Access``. We can keep them separate (and assume access is a subset of ``Tenant``), or find a way to blend them + permissions. The latter is nice because then we have a singular "what permission does this actor have", but will require some refactoring.

dcramer commented on src/sentry/db/models/manager.py:
```diff
@@ -313,3 +321,27 @@ def post_delete(self, instance, **kwargs):
+    # TODO(dcramer): its not clear to me that we actually need to
+    # use BaseManager ever on related_fields (other this changes
+    # that behavior)
+    use_for_related_fields = False
```
One concern here is we might actually want this in some cases. We wouldn't want it ever for ForeignKey's, but for a reverse mapper we would (e.g. organization.project_set should be constrained on team_id eventually).

dcramer replied:
related_name is actually a great answer since we can control it via mixins. going that route.

dcramer commented on src/sentry/db/models/mixins.py:
```diff
@@ -0,0 +1,12 @@
+
+
+class OrganizationBoundMixin(object):
+    organization = FlexibleForeignKey('sentry.Organization')
```
likely will need to tweak this given some models use an integer and some use a foreignkey

might just create two diff mixins for it

dcramer commented on src/sentry/db/models/queryset.py:
```diff
@@ -0,0 +1,102 @@
+            rv._binding_criteria_fn = self._binding_criteria_fn
+            return rv
+
+        # XXX(dcramer): if we could find a way to clear these parameters (which is tricky)
```
i kept this comment in case anyone ever looks at this code and asks "why cant we use _setup_query"

dcramer commented on src/sentry/db/models/queryset.py:
```diff
@@ -0,0 +1,102 @@
+                qs = self._clone_with_constraints()
+            return cls.count(qs)
+
+        def earliest(self, *args, **kwargs):
```
we dont use this and i didnt really care to implement them

dcramer commented on src/sentry/middleware/env.py:
```diff
@@ -9,9 +9,11 @@ class SentryEnvMiddleware(object):
     def process_request(self, request):
         # bind request to env
         env.request = request
+        env.tenant = None
```
need to ensure celery tasks are wired to ensure tenant is cleared on each task run (I forget if it hits our request_finished signal)

dcramer replied:
done

dcramer commented:
also still need to handle superuser path (though arguably that should be explicit per-query)

dcramer commented:
http://www.thesaurus.com/browse/omnipotent

dcramer commented:
- renamed ``unconstrained_unsafe`` to ``unrestricted_unsafe``
- added tenants.UnrestrictedTenant
- add TestCase.tenant(tenant) context manager
- add TestCase.unrestricted_tenant() context manager

dcramer commented:
Open question: should Celery tasks be set to Unrestricted by default? I can't think of any task that requires verification of scoping.

mitsuhiko commented:
@dcramer do you still want to go ahead with this?

dcramer commented:
How do you all still feel about this. I think its still a good idea to protect ourselves, but it does have some minor perf implications.

dcramer commented on src/sentry/tasks/auto_resolve_issues.py:
```diff
@@ -50,7 +50,7 @@ def schedule_auto_resolution():
                    soft_time_limit=60)
 def auto_resolve_project_issues(project_id, cutoff=None, chunk_size=1000,
                                 **kwargs):
-    project = Project.objects.get_from_cache(id=project_id)
```
should probably bind tenants in tasks in the same vein as zeus does

e.g. it grabs the first object (unrestricted), and then scopes a tenant based on that objects access pattern

dcramer commented:
Talked in Slack that we're all in agreement that this is a good direction.

I will:

- rebase
- add basic tests proving all of the base models function as expected
- [likey] change tasks to bind tenants

EvanPurkhiser commented:
Removing platform team on review, since this is stale at the moment.
